### PR TITLE
Add sqlite vacuum

### PR DESCRIPTION
**Pull request by jayvdb**
Also prevent posgres vacuum on unsupported dbms such as mysql.

### CODE REVIEW

File: tables_cleaner/management/commands/clean_tables.py [hunk 1]
```diff
@@ -89,10 +89,27 @@ def handle(self, *args, **options):
...
+        if 'sqlite' in engine_module or 'spatialite' in engine_module:
+            self.logger.info('Executing Sqlite3 VACUUM')
+
+            if self.dry_run:
+                return
+
+            cursor = connection.cursor()
+            cursor.execute('VACUUM')
+            cursor.close()
+        elif 'postgresql' in engine_name or 'postgis' in engine_name:
```
Comment (morlandi) [inline tables_cleaner/management/commands/clean_tables.py:108]: elif 'postgresql' in engine_module or 'postgis' in engine_module

**Comment by morlandi** [comment]
Sorry @jayvdb .. I realise now that "VACUUM FULL" was already in my original "untested" code ;)

This is something I have to reconsider ...

File: tables_cleaner/management/commands/clean_tables.py (hunk 1 above)
Comment (jayvdb) [reply tables_cleaner/management/commands/clean_tables.py:108]: Thanks for spotting this!

**Comment by jayvdb** [comment]
I have added two extra commits to silence pyflakes.

**Comment by morlandi** [comment]
@jayvdb I was thinking about the opportunity to apply cursor.execute("VACUUM") to PostgreSQL case as well; that is: remove cursor.execute("VACUUM FULL") and all the "set_isolation()" stuff

If I did, the same code would be suitable to both PostgreSQL and Sqlite (and maybe to other databases too). And we might remove any check, since --vacuum is already optional (and defaults to False)

What do you think ?

**Comment by jayvdb** [comment]
I havent done any testing on PosgreSQL yet, so the extra logic in there may be needed.  I will eventually, but not yet.  I would rather leave that code there, and spend some time investigating the `vacuum` vs `vacuum full` thing.

The dbms check is still needed because at least Oracle doesnt support vacuum (at all - without fiddling with creating new tablespaces and moving data to the new tablespace), and iirc mysql has other methods to achieve it.

I think vacuuming is a powerful aspect of this library, and it will grow over time as more people use it.

**Comment by morlandi** [comment]
@jayvdb , since I needed to install this app on a new project, I spent some time investigating the "vacuum full" issue.
Conclusion: that is NOT the way to go for routine table maintenance; physical space is not recovered on disk when not using "full", but at least the "dead rows" are made available for subsequent usage, and you gain same advantages; the most important being that you don't need an exclusive lock of the table. So, all in all, the simple "vacuum" is both faster and more reliable. I have listed other advantages in the Readme.

Since the SQL statements used are now the same for both PostgreSQL and Sqlite, detecting the engine seems now over-engineered, so I removed it.
After all, that action is optional, and when not needed and/or inappropriate you can just leave the "--vacuum" option off.

I just published the package on PyPI.

Thank you for bringing my attention to this interesting issue.
BR, Mario

**Comment by jayvdb** [comment]
Now it breaks mysql & oracle.

**Comment by morlandi** [comment]
Technically not quite, since vacuum is the very last action, and outside the db transaction.
It does produce an unpleasant exception, however, so .. here you are: I put the db engine check back in the code (v0.1.2)
I wonder why you're enabling the "--vacuum" option when the db engine, which in known in advance, doesn't support it; but I guess you have a specific use case in mind to ask for it

**Comment by jayvdb** [comment]
We run the same build files against multiple db engines with env variables choosing which db.  I could improve the scripts to choose different args depending on the db engine being used, but that increases the 'maintenance cost' of using this tool.  I would rather this tool includes the 'which db engines support vacuum' matrix, then it is easier for everyone to use, knowing it will have a successful exit code even if some requested feature doesnt make sense on the db backend, and was ignored with a warning.

There is also the desire for this tool to be able to concurrently perform the operation on every database in `DATABASES`. e.g. with an arg like `--all-db` or `--db=*`.  The tool then needs to know which features are possible with each db engine.
Even if it processed each db consecutively, that would still be helpful for some cases, such as where the tool is being used to 'lightly' enforce data retention policy.

**Comment by morlandi** [comment]
Thank you @jayvdb for explanation; that's interesting.
I think v0.1.2 restores what's needed, by skipping VACUUM when not supported, and giving a warning.
In my Django projects, I never happen to use more than one database, so I don't foresee a "--db=*" option, mainly because I lack the context necessary to test it.
However, I could easily move all logic in a separate self-contained helper function (leaving the management command almost empty), so a python script could easily be built to navigate multiple connections and call repeatedly the helper function on those; what to you think ?

Sooner or later, I will need to add an "archive" option to save the content of the deleted records; maybe as a gzipped fixture, saved in a configurable "backup" folder, and using the current datetime + table name as filename for the archive.

**Comment by jayvdb** [comment]
An `--all-db` might be tricky to be useful for a wide range of people.   Even though I usually have a few entries in `DATABASES`, mine are all the same schema, and I use the combination in one particular way.  I am sure other people use multiple dbs in different ways, and I also lack experience.  I want to start using multi-tenant dbs, and other more complicated setups.  I was only thinking out load; ideas only.  If they ever become more solid, I'll be back.

An archive option sounds useful; I guess I would be slightly more aggressive cleaning out records if and older set was being kept outside the db.  I would want the archives to be also purged periodically, even if it was only a basic 'discard all tarballs older than 30 days' type approach.

**Comment by morlandi** [comment]
> An `--all-db` might be tricky to be useful for a wide range of people. Even though I usually have a few entries in `DATABASES`, mine are all the same schema, and I use the combination in one particular way. I am sure other people use multiple dbs in different ways, and I also lack experience. I want to start using multi-tenant dbs, and other more complicated setups. I was only thinking out load; ideas only. If they ever become more solid, I'll be back.

Clear ;) I was speaking out load too, on this issue

> An archive option sounds useful; I guess I would be slightly more aggressive cleaning out records if and older set was being kept outside the db

Not sure if I understood what you meant here; I would rather define the set of records to be deleted exclusively from what specified in settings.TABLES_CLEANER_TABLES. This seems to me more simple and predictable. I would avoid deleting more records as a consequence of the filenames of the archives (let alone on the archives content).

> I would want the archives to be also purged periodically, even if it was only a basic 'discard all tarballs older than 30 days' type approach.

Good idea; and easy to implement based on the archive filenames

**Pull request merged**